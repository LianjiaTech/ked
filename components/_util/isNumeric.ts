const isNumeric = (value: any): boolean => {
  return !Number.isNaN(parseFloat(value)) && Number.isFinite(value);
};

export default isNumeric;
