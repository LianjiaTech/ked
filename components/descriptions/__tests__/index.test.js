import React from 'react';
import MockDate from 'mockdate';
import { mount } from 'enzyme';
import Descriptions from '..';
import mountTest from '../../../../tests/shared/mountTest';
import { resetWarned } from '../../_util/warning';

jest.mock('enquire.js', () => {
  let that;
  let unmatchFun;
  return {
    unregister: jest.fn(),
    register: (media, options) => {
      if (media === '(max-width: 575px)') {
        that = this;
        options.match.call(that);
        unmatchFun = options.unmatch;
      }
    },
    callunmatch() {
      unmatchFun.call(that);
    },
  };
});

describe('Descriptions', () => {
  mountTest(Descriptions);

  const errorSpy = jest.spyOn(console, 'error').mockImplementation(() => {});

  afterEach(() => {
    MockDate.reset();
    errorSpy.mockReset();
  });

  afterAll(() => {
    errorSpy.mockRestore();
  });

  it('when max-width: 575px，column=1', () => {
    // eslint-disable-next-line global-require
    const enquire = require('enquire.js');
    const wrapper = mount(
      <Descriptions>
        <Descriptions.Item label="Product">Cloud Database</Descriptions.Item>
        <Descriptions.Item label="Billing">Prepaid</Descriptions.Item>
        <Descriptions.Item label="time">18:00:00</Descriptions.Item>
        <Descriptions.Item label="Amount">$80.00</Descriptions.Item>
        <Descriptions.Item>No-Label</Descriptions.Item>
      </Descriptions>,
    );
    expect(wrapper.find('tr')).toHaveLength(5);
    expect(wrapper.find('.ant-descriptions-item-no-label')).toHaveLength(1);

    enquire.callunmatch();
    wrapper.unmount();
  });

  it('when max-width: 575px，column=2', () => {
    // eslint-disable-next-line global-require
    const enquire = require('enquire.js');
    const wrapper = mount(
      <Descriptions column={{ xs: 2 }}>
        <Descriptions.Item label="Product">Cloud Database</Descriptions.Item>
        <Descriptions.Item label="Billing">Prepaid</Descriptions.Item>
        <Descriptions.Item label="time">18:00:00</Descriptions.Item>
        <Descriptions.Item label="Amount">$80.00</Descriptions.Item>
      </Descriptions>,
    );
    expect(wrapper.find('tr')).toHaveLength(2);

    enquire.callunmatch();
    wrapper.unmount();
  });

  it('column is number', () => {
    // eslint-disable-next-line global-require
    const wrapper = mount(
      <Descriptions column="3">
        <Descriptions.Item label="Product">Cloud Database</Descriptions.Item>
        <Descriptions.Item label="Billing">Prepaid</Descriptions.Item>
        <Descriptions.Item label="time">18:00:00</Descriptions.Item>
        <Descriptions.Item label="Amount">$80.00</Descriptions.Item>
      </Descriptions>,
    );
    expect(wrapper).toMatchSnapshot();
    wrapper.unmount();
  });

  it('when typeof column is object', () => {
    const wrapper = mount(
      <Descriptions column={{ xs: 8, sm: 16, md: 24 }}>
        <Descriptions.Item label="Product">Cloud Database</Descriptions.Item>
        <Descriptions.Item label="Billing">Prepaid</Descriptions.Item>
        <Descriptions.Item label="time">18:00:00</Descriptions.Item>
        <Descriptions.Item label="Amount">$80.00</Descriptions.Item>
      </Descriptions>,
    );
    expect(wrapper.instance().getColumn()).toBe(8);
    wrapper.unmount();
  });

  it('warning if ecceed the row span', () => {
    resetWarned();

    mount(
      <Descriptions column={3}>
        <Descriptions.Item label="Product" span={2}>
          Cloud Database
        </Descriptions.Item>
        <Descriptions.Item label="Billing" span={2}>
          Prepaid
        </Descriptions.Item>
      </Descriptions>,
    );
    expect(errorSpy).toHaveBeenCalledWith(
      'Warning: [ked: Descriptions] Sum of column `span` in a line exceeds `column` of Descriptions.',
    );
  });

  it('when item is rendered conditionally', () => {
    const hasDiscount = false;
    const wrapper = mount(
      <Descriptions>
        <Descriptions.Item label="Product">Cloud Database</Descriptions.Item>
        <Descriptions.Item label="Billing">Prepaid</Descriptions.Item>
        <Descriptions.Item label="time">18:00:00</Descriptions.Item>
        <Descriptions.Item label="Amount">$80.00</Descriptions.Item>
        {hasDiscount && <Descriptions.Item label="Discount">$20.00</Descriptions.Item>}
      </Descriptions>,
    );
    expect(wrapper).toMatchSnapshot();
    wrapper.unmount();
  });

  it('vertical layout', () => {
    // eslint-disable-next-line global-require
    const wrapper = mount(
      <Descriptions layout="vertical">
        <Descriptions.Item label="Product">Cloud Database</Descriptions.Item>
        <Descriptions.Item label="Billing">Prepaid</Descriptions.Item>
        <Descriptions.Item label="time">18:00:00</Descriptions.Item>
        <Descriptions.Item label="Amount">$80.00</Descriptions.Item>
      </Descriptions>,
    );
    expect(wrapper).toMatchSnapshot();
    wrapper.unmount();
  });

  it('Descriptions.Item support className', () => {
    const wrapper = mount(
      <Descriptions>
        <Descriptions.Item label="Product" className="my-class">
          Cloud Database
        </Descriptions.Item>
      </Descriptions>,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('Descriptions support colon', () => {
    const wrapper = mount(
      <Descriptions colon={false}>
        <Descriptions.Item label="Product">Cloud Database</Descriptions.Item>
      </Descriptions>,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('Descriptions support style', () => {
    const wrapper = mount(
      <Descriptions style={{ backgroundColor: '#e8e8e8' }}>
        <Descriptions.Item>Cloud Database</Descriptions.Item>
      </Descriptions>,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('keep key', () => {
    const wrapper = mount(
      <Descriptions>
        <Descriptions.Item key="bamboo" />
      </Descriptions>,
    );

    expect(wrapper.find('Col').key()).toBe('label-bamboo');
  });

  // https://github.com/ant-design/ant-design/issues/19887
  it('should work with React Fragment', () => {
    if (!React.Fragment) {
      return;
    }
    const wrapper = mount(
      <Descriptions>
        <Descriptions.Item label="bamboo">bamboo</Descriptions.Item>
        <>
          <Descriptions.Item label="bamboo">bamboo</Descriptions.Item>
          <Descriptions.Item label="bamboo">bamboo</Descriptions.Item>
        </>
      </Descriptions>,
    );

    expect(wrapper.render()).toMatchSnapshot();
  });
});
