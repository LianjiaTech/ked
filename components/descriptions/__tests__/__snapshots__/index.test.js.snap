// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Descriptions Descriptions support colon 1`] = `
<Descriptions
  colon={false}
  column={
    Object {
      "lg": 3,
      "md": 3,
      "sm": 2,
      "xl": 3,
      "xs": 1,
      "xxl": 3,
    }
  }
  size="default"
>
  <div
    className="ant-descriptions"
  >
    <div
      className="ant-descriptions-view"
    >
      <table>
        <tbody>
          <tr
            className="ant-descriptions-row"
            key="0"
          >
            <Col
              bordered={false}
              child={
                <DescriptionsItem
                  label="Product"
                  prefixCls="ant-descriptions"
                  span={1}
                >
                  Cloud Database
                </DescriptionsItem>
              }
              colon={false}
              key="label-0"
              layout="horizontal"
              type="label"
            >
              <td
                className="ant-descriptions-item"
                colSpan={1}
              >
                <span
                  className="ant-descriptions-item-label"
                  key="label"
                >
                  Product
                </span>
                <span
                  className="ant-descriptions-item-content"
                  key="content"
                >
                  Cloud Database
                </span>
              </td>
            </Col>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</Descriptions>
`;

exports[`Descriptions Descriptions support style 1`] = `
<Descriptions
  column={
    Object {
      "lg": 3,
      "md": 3,
      "sm": 2,
      "xl": 3,
      "xs": 1,
      "xxl": 3,
    }
  }
  size="default"
  style={
    Object {
      "backgroundColor": "#e8e8e8",
    }
  }
>
  <div
    className="ant-descriptions"
    style={
      Object {
        "backgroundColor": "#e8e8e8",
      }
    }
  >
    <div
      className="ant-descriptions-view"
    >
      <table>
        <tbody>
          <tr
            className="ant-descriptions-row"
            key="0"
          >
            <Col
              bordered={false}
              child={
                <DescriptionsItem
                  prefixCls="ant-descriptions"
                  span={1}
                >
                  Cloud Database
                </DescriptionsItem>
              }
              colon={true}
              key="label-0"
              layout="horizontal"
              type="label"
            >
              <td
                className="ant-descriptions-item"
                colSpan={1}
              >
                <span
                  className="ant-descriptions-item-label ant-descriptions-item-colon ant-descriptions-item-no-label"
                  key="label"
                />
                <span
                  className="ant-descriptions-item-content"
                  key="content"
                >
                  Cloud Database
                </span>
              </td>
            </Col>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</Descriptions>
`;

exports[`Descriptions Descriptions.Item support className 1`] = `
<Descriptions
  column={
    Object {
      "lg": 3,
      "md": 3,
      "sm": 2,
      "xl": 3,
      "xs": 1,
      "xxl": 3,
    }
  }
  size="default"
>
  <div
    className="ant-descriptions"
  >
    <div
      className="ant-descriptions-view"
    >
      <table>
        <tbody>
          <tr
            className="ant-descriptions-row"
            key="0"
          >
            <Col
              bordered={false}
              child={
                <DescriptionsItem
                  className="my-class"
                  label="Product"
                  prefixCls="ant-descriptions"
                  span={1}
                >
                  Cloud Database
                </DescriptionsItem>
              }
              colon={true}
              key="label-0"
              layout="horizontal"
              type="label"
            >
              <td
                className="ant-descriptions-item my-class"
                colSpan={1}
              >
                <span
                  className="ant-descriptions-item-label ant-descriptions-item-colon"
                  key="label"
                >
                  Product
                </span>
                <span
                  className="ant-descriptions-item-content"
                  key="content"
                >
                  Cloud Database
                </span>
              </td>
            </Col>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</Descriptions>
`;

exports[`Descriptions column is number 1`] = `
<Descriptions
  column="3"
  size="default"
>
  <div
    className="ant-descriptions"
  >
    <div
      className="ant-descriptions-view"
    >
      <table>
        <tbody>
          <tr
            className="ant-descriptions-row"
            key="0"
          >
            <Col
              bordered={false}
              child={
                <DescriptionsItem
                  label="Product"
                  prefixCls="ant-descriptions"
                >
                  Cloud Database
                </DescriptionsItem>
              }
              colon={true}
              key="label-0"
              layout="horizontal"
              type="label"
            >
              <td
                className="ant-descriptions-item"
                colSpan={1}
              >
                <span
                  className="ant-descriptions-item-label ant-descriptions-item-colon"
                  key="label"
                >
                  Product
                </span>
                <span
                  className="ant-descriptions-item-content"
                  key="content"
                >
                  Cloud Database
                </span>
              </td>
            </Col>
            <Col
              bordered={false}
              child={
                <DescriptionsItem
                  label="Billing"
                  prefixCls="ant-descriptions"
                >
                  Prepaid
                </DescriptionsItem>
              }
              colon={true}
              key="label-1"
              layout="horizontal"
              type="label"
            >
              <td
                className="ant-descriptions-item"
                colSpan={1}
              >
                <span
                  className="ant-descriptions-item-label ant-descriptions-item-colon"
                  key="label"
                >
                  Billing
                </span>
                <span
                  className="ant-descriptions-item-content"
                  key="content"
                >
                  Prepaid
                </span>
              </td>
            </Col>
            <Col
              bordered={false}
              child={
                <DescriptionsItem
                  label="time"
                  prefixCls="ant-descriptions"
                >
                  18:00:00
                </DescriptionsItem>
              }
              colon={true}
              key="label-2"
              layout="horizontal"
              type="label"
            >
              <td
                className="ant-descriptions-item"
                colSpan={1}
              >
                <span
                  className="ant-descriptions-item-label ant-descriptions-item-colon"
                  key="label"
                >
                  time
                </span>
                <span
                  className="ant-descriptions-item-content"
                  key="content"
                >
                  18:00:00
                </span>
              </td>
            </Col>
          </tr>
          <tr
            className="ant-descriptions-row"
            key="1"
          >
            <Col
              bordered={false}
              child={
                <DescriptionsItem
                  label="Amount"
                  prefixCls="ant-descriptions"
                  span={3}
                >
                  $80.00
                </DescriptionsItem>
              }
              colon={true}
              key="label-0"
              layout="horizontal"
              type="label"
            >
              <td
                className="ant-descriptions-item"
                colSpan={3}
              >
                <span
                  className="ant-descriptions-item-label ant-descriptions-item-colon"
                  key="label"
                >
                  Amount
                </span>
                <span
                  className="ant-descriptions-item-content"
                  key="content"
                >
                  $80.00
                </span>
              </td>
            </Col>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</Descriptions>
`;

exports[`Descriptions should work with React Fragment 1`] = `
<div
  class="ant-descriptions"
>
  <div
    class="ant-descriptions-view"
  >
    <table>
      <tbody>
        <tr
          class="ant-descriptions-row"
        >
          <td
            class="ant-descriptions-item"
            colspan="1"
          >
            <span
              class="ant-descriptions-item-label ant-descriptions-item-colon"
            >
              bamboo
            </span>
            <span
              class="ant-descriptions-item-content"
            >
              bamboo
            </span>
          </td>
        </tr>
        <tr
          class="ant-descriptions-row"
        >
          <td
            class="ant-descriptions-item"
            colspan="1"
          >
            <span
              class="ant-descriptions-item-label ant-descriptions-item-colon"
            >
              bamboo
            </span>
            <span
              class="ant-descriptions-item-content"
            >
              bamboo
            </span>
          </td>
        </tr>
        <tr
          class="ant-descriptions-row"
        >
          <td
            class="ant-descriptions-item"
            colspan="1"
          >
            <span
              class="ant-descriptions-item-label ant-descriptions-item-colon"
            >
              bamboo
            </span>
            <span
              class="ant-descriptions-item-content"
            >
              bamboo
            </span>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`Descriptions vertical layout 1`] = `
<Descriptions
  column={
    Object {
      "lg": 3,
      "md": 3,
      "sm": 2,
      "xl": 3,
      "xs": 1,
      "xxl": 3,
    }
  }
  layout="vertical"
  size="default"
>
  <div
    className="ant-descriptions"
  >
    <div
      className="ant-descriptions-view"
    >
      <table>
        <tbody>
          <tr
            className="ant-descriptions-row"
            key="label-0"
          >
            <Col
              bordered={false}
              child={
                <DescriptionsItem
                  label="Product"
                  prefixCls="ant-descriptions"
                >
                  Cloud Database
                </DescriptionsItem>
              }
              colon={true}
              key="label-0"
              layout="vertical"
              type="label"
            >
              <td
                className="ant-descriptions-item"
                colSpan={1}
              >
                <span
                  className="ant-descriptions-item-label ant-descriptions-item-colon"
                  key="label"
                >
                  Product
                </span>
              </td>
            </Col>
          </tr>
          <tr
            className="ant-descriptions-row"
            key="content-0"
          >
            <Col
              bordered={false}
              child={
                <DescriptionsItem
                  label="Product"
                  prefixCls="ant-descriptions"
                >
                  Cloud Database
                </DescriptionsItem>
              }
              colon={true}
              key="content-0"
              layout="vertical"
              type="content"
            >
              <td
                className="ant-descriptions-item"
                colSpan={1}
              >
                <span
                  className="ant-descriptions-item-content"
                  key="content"
                >
                  Cloud Database
                </span>
              </td>
            </Col>
          </tr>
          <tr
            className="ant-descriptions-row"
            key="label-1"
          >
            <Col
              bordered={false}
              child={
                <DescriptionsItem
                  label="Billing"
                  prefixCls="ant-descriptions"
                >
                  Prepaid
                </DescriptionsItem>
              }
              colon={true}
              key="label-0"
              layout="vertical"
              type="label"
            >
              <td
                className="ant-descriptions-item"
                colSpan={1}
              >
                <span
                  className="ant-descriptions-item-label ant-descriptions-item-colon"
                  key="label"
                >
                  Billing
                </span>
              </td>
            </Col>
          </tr>
          <tr
            className="ant-descriptions-row"
            key="content-1"
          >
            <Col
              bordered={false}
              child={
                <DescriptionsItem
                  label="Billing"
                  prefixCls="ant-descriptions"
                >
                  Prepaid
                </DescriptionsItem>
              }
              colon={true}
              key="content-0"
              layout="vertical"
              type="content"
            >
              <td
                className="ant-descriptions-item"
                colSpan={1}
              >
                <span
                  className="ant-descriptions-item-content"
                  key="content"
                >
                  Prepaid
                </span>
              </td>
            </Col>
          </tr>
          <tr
            className="ant-descriptions-row"
            key="label-2"
          >
            <Col
              bordered={false}
              child={
                <DescriptionsItem
                  label="time"
                  prefixCls="ant-descriptions"
                >
                  18:00:00
                </DescriptionsItem>
              }
              colon={true}
              key="label-0"
              layout="vertical"
              type="label"
            >
              <td
                className="ant-descriptions-item"
                colSpan={1}
              >
                <span
                  className="ant-descriptions-item-label ant-descriptions-item-colon"
                  key="label"
                >
                  time
                </span>
              </td>
            </Col>
          </tr>
          <tr
            className="ant-descriptions-row"
            key="content-2"
          >
            <Col
              bordered={false}
              child={
                <DescriptionsItem
                  label="time"
                  prefixCls="ant-descriptions"
                >
                  18:00:00
                </DescriptionsItem>
              }
              colon={true}
              key="content-0"
              layout="vertical"
              type="content"
            >
              <td
                className="ant-descriptions-item"
                colSpan={1}
              >
                <span
                  className="ant-descriptions-item-content"
                  key="content"
                >
                  18:00:00
                </span>
              </td>
            </Col>
          </tr>
          <tr
            className="ant-descriptions-row"
            key="label-3"
          >
            <Col
              bordered={false}
              child={
                <DescriptionsItem
                  label="Amount"
                  prefixCls="ant-descriptions"
                  span={1}
                >
                  $80.00
                </DescriptionsItem>
              }
              colon={true}
              key="label-0"
              layout="vertical"
              type="label"
            >
              <td
                className="ant-descriptions-item"
                colSpan={1}
              >
                <span
                  className="ant-descriptions-item-label ant-descriptions-item-colon"
                  key="label"
                >
                  Amount
                </span>
              </td>
            </Col>
          </tr>
          <tr
            className="ant-descriptions-row"
            key="content-3"
          >
            <Col
              bordered={false}
              child={
                <DescriptionsItem
                  label="Amount"
                  prefixCls="ant-descriptions"
                  span={1}
                >
                  $80.00
                </DescriptionsItem>
              }
              colon={true}
              key="content-0"
              layout="vertical"
              type="content"
            >
              <td
                className="ant-descriptions-item"
                colSpan={1}
              >
                <span
                  className="ant-descriptions-item-content"
                  key="content"
                >
                  $80.00
                </span>
              </td>
            </Col>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</Descriptions>
`;

exports[`Descriptions when item is rendered conditionally 1`] = `
<Descriptions
  column={
    Object {
      "lg": 3,
      "md": 3,
      "sm": 2,
      "xl": 3,
      "xs": 1,
      "xxl": 3,
    }
  }
  size="default"
>
  <div
    className="ant-descriptions"
  >
    <div
      className="ant-descriptions-view"
    >
      <table>
        <tbody>
          <tr
            className="ant-descriptions-row"
            key="0"
          >
            <Col
              bordered={false}
              child={
                <DescriptionsItem
                  label="Product"
                  prefixCls="ant-descriptions"
                >
                  Cloud Database
                </DescriptionsItem>
              }
              colon={true}
              key="label-0"
              layout="horizontal"
              type="label"
            >
              <td
                className="ant-descriptions-item"
                colSpan={1}
              >
                <span
                  className="ant-descriptions-item-label ant-descriptions-item-colon"
                  key="label"
                >
                  Product
                </span>
                <span
                  className="ant-descriptions-item-content"
                  key="content"
                >
                  Cloud Database
                </span>
              </td>
            </Col>
          </tr>
          <tr
            className="ant-descriptions-row"
            key="1"
          >
            <Col
              bordered={false}
              child={
                <DescriptionsItem
                  label="Billing"
                  prefixCls="ant-descriptions"
                >
                  Prepaid
                </DescriptionsItem>
              }
              colon={true}
              key="label-0"
              layout="horizontal"
              type="label"
            >
              <td
                className="ant-descriptions-item"
                colSpan={1}
              >
                <span
                  className="ant-descriptions-item-label ant-descriptions-item-colon"
                  key="label"
                >
                  Billing
                </span>
                <span
                  className="ant-descriptions-item-content"
                  key="content"
                >
                  Prepaid
                </span>
              </td>
            </Col>
          </tr>
          <tr
            className="ant-descriptions-row"
            key="2"
          >
            <Col
              bordered={false}
              child={
                <DescriptionsItem
                  label="time"
                  prefixCls="ant-descriptions"
                >
                  18:00:00
                </DescriptionsItem>
              }
              colon={true}
              key="label-0"
              layout="horizontal"
              type="label"
            >
              <td
                className="ant-descriptions-item"
                colSpan={1}
              >
                <span
                  className="ant-descriptions-item-label ant-descriptions-item-colon"
                  key="label"
                >
                  time
                </span>
                <span
                  className="ant-descriptions-item-content"
                  key="content"
                >
                  18:00:00
                </span>
              </td>
            </Col>
          </tr>
          <tr
            className="ant-descriptions-row"
            key="3"
          >
            <Col
              bordered={false}
              child={
                <DescriptionsItem
                  label="Amount"
                  prefixCls="ant-descriptions"
                  span={1}
                >
                  $80.00
                </DescriptionsItem>
              }
              colon={true}
              key="label-0"
              layout="horizontal"
              type="label"
            >
              <td
                className="ant-descriptions-item"
                colSpan={1}
              >
                <span
                  className="ant-descriptions-item-label ant-descriptions-item-colon"
                  key="label"
                >
                  Amount
                </span>
                <span
                  className="ant-descriptions-item-content"
                  key="content"
                >
                  $80.00
                </span>
              </td>
            </Col>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</Descriptions>
`;
